src/foam/dao/AbstractF3FileJournal.js

```diff
@@ -0,0 +1,485 @@
+          }
+        }
+
+        public void endJob() {
+          if ( foam.util.SafetyUtil.isEmpty(record_) ) return;
+
+          try {
+            writeComment_(x, obj);
+            writeRemove_(x, record_, foam.util.SafetyUtil.isEmpty(prefix) ? "" : prefix + ".");
+
+            if ( isLast() ) getWriter().flush();
+          } catch (Throwable t) {
+            getLogger().error("Failed to write put entry to journal", t);
+          }
+        }
+      });
+
+      return obj;
+      `
+    },
+    {
+      name: 'writeRemove_',
+      javaThrows: [
+        'java.io.IOException'
+      ],
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'record',
+          type: 'String'
+        },
+        {
+          name: 'prefix',
+          type: 'String'
+        }
+      ],
+      javaCode: `
+        write_(sb.get()
+          .append(prefix)
+          .append("r(")
+          .append(record)
+          .append(")")
+          .toString());
+      `
+    },
+    {
+      name: 'write_',
+     // synchronized: true,
+      javaThrows: [
+        'java.io.IOException'
+      ],
+      args: [
+        {
+          class: 'String',
+          name: 'data'
+        }
+      ],
+      javaCode: `
+        BufferedWriter writer = getWriter();
+        writer.write(data);
+        writer.newLine();
+      `
+    },
+    {
+      name: 'writeComment_',
+     // synchronized: true,
+      javaThrows: [
+        'java.io.IOException'
+      ],
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'obj',
+          type: 'foam.core.FObject'
+        }
+      ],
+      javaCode: `
+        User user = ((Subject) x.get("subject")).getUser();
+        if ( user == null || user.getId() <= 1 ) return;
+        if ( obj instanceof LastModifiedByAware && ((LastModifiedByAware) obj).getLastModifiedBy() != 0L ) return;
+
+        write_(sb.get()
+          .append("// Modified by ")
+          .append(user.toSummary())
+          .append(" (")
+          .append(user.getId())
+          .append(") at ")
+          .append(getTimeStamper().createTimestamp())
+          .toString());
+      `
+    },
+    {
+      name: 'getEntry',
+      documentation: 'retrieves ameaningful unit of text from the journal',
+      type: 'String',
+      args: [
+        {
+          name: 'reader',
+          type: 'BufferedReader'
+        }
+      ],
+      javaCode: `
+        try {
+          String line = reader.readLine();
+          if ( line == null ) return null;
+          if ( ! line.equals("p({") && ! line.equals("r({") ) return line;
+          StringBuilder sb = new StringBuilder();
+          sb.append(line);
+          while( ! line.equals("})") ) {
+            if ( (line = reader.readLine()) == null ) break;
+            sb.append("\\n");
+            sb.append(line);
+          }
+          return sb.toString().trim();
```

Comment:
We should make the StringPStream take a CharSequence instead. Then we don't need to call toString() here.